File: src/main/java/ru/serdtsev/homemoney/account/Reserve.java
package ru.serdtsev.homemoney.account;

import ru.serdtsev.homemoney.balancesheet.BalanceSheet;
import ru.serdtsev.homemoney.dao.MoneyTrnsDao;
import ru.serdtsev.homemoney.utils.Utils;

import javax.persistence.DiscriminatorValue;
import javax.persistence.Entity;
import javax.persistence.Table;
import java.math.BigDecimal;
import java.sql.Date;

@Entity
@Table(name= "reserves")
@DiscriminatorValue("reserve")
public class Reserve extends Balance {
  private BigDecimal target;

  @SuppressWarnings({"unused", "WeakerAccess"})
  protected Reserve() {
  }

  public Reserve(BalanceSheet balanceSheet, String name, Date created, Boolean isArc, String currencyCode,
      BigDecimal value, BigDecimal target) {
    super(balanceSheet, AccountType.reserve, name, created, isArc, currencyCode, value, null);
    this.target = target;
  }

  public void init() {
    super.init(null);
    target = Utils.nvl(target, BigDecimal.ZERO);
  }

  public void merge(Reserve reserve, ReserveRepository reserveRepo, MoneyTrnsDao moneyTrnsDao) {
    super.merge(reserve, reserveRepo, moneyTrnsDao);
    setTarget(reserve.getTarget());
  }

  public BigDecimal getTarget() {
    return target;
  }

  public void setTarget(BigDecimal target) {
    this.target = target;
  }
}

File: src/main/kotlin/ru/serdtsev/homemoney/domain/model/account/Reserve.kt
package ru.serdtsev.homemoney.domain.model.account

import ru.serdtsev.homemoney.domain.event.DomainEventPublisher
import ru.serdtsev.homemoney.infra.ApiRequestContextHolder
import java.math.BigDecimal
import java.time.LocalDate
import java.util.*

open class Reserve(
    id: UUID,
    name: String,
    createdDate: LocalDate = LocalDate.now(),
    currencyCode: String = ApiRequestContextHolder.balanceSheet.currencyCode,
    value: BigDecimal = BigDecimal.ZERO,
    open var target: BigDecimal = BigDecimal.ZERO,
    isArc: Boolean = false
) : Balance(id, AccountType.reserve, name, createdDate, isArc, currencyCode, value) {
    internal constructor(name: String, value: BigDecimal = BigDecimal.ZERO, target: BigDecimal = BigDecimal.ZERO) :
            this(UUID.randomUUID(), name, value = value, target = target)

    companion object {
        fun merge(from: Reserve, to: Reserve) {
            Balance.merge(from, to)
            to.target = from.target
            DomainEventPublisher.instance.publish(to)
        }
    }
}